pkg/register/component.go

```diff
@@ -0,0 +1,227 @@
+		UserID:               &userID,
+		Gender:               user.Gender,
+		BirthDate:            user.BirthDate,
+		BirthLocation:        user.BirthLocation,
+		IDDocumentType:       user.IDDocumentType,
+		IDDocumentNumber:     user.IDDocumentNumber,
+		IDDocumentExpiration: user.IDDocumentExpiration,
+	})
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't store user details in database", "err", err.Error())
+		return "", err
+	}
+
+	// Send execute actions email
+	if realmConf.RegisterRequiredActions != nil && len(*realmConf.RegisterRequiredActions) > 0 {
+		err = c.keycloakClient.ExecuteActionsEmail(accessToken, "master", userID, *realmConf.RegisterRequiredActions)
+		if err != nil {
+			c.logger.Warn(ctx, "msg", "ExecuteActionsEmail failed", "err", err.Error())
+			return "", err
+		}
+	}
+
+	// Commit the DB transaction
+	err = tx.Commit()
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't commit transaction", "err", err.Error())
+		return "", errorhandler.CreateInternalServerError("database")
+	}
+
+	// store the API call into the DB
+	c.reportEvent(ctx, "REGISTER_USER", database.CtEventRealmName, realmName, database.CtEventUserID, userID)
+
+	return username, nil
+}
+
+func (c *component) checkExistingUser(ctx context.Context, accessToken, realmName string, user apiregister.User) error {
+	// Email already used?
+	var existingUser, err = c.usersDBModule.GetUser(ctx, realmName, nil, nil, user.EmailAddress)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't get user from db", "err", err.Error())
+		return errorhandler.CreateInternalServerError("database")
+	}
+	if existingUser == nil {
+		// Free to go on processing this user creation
+		return nil
+	}
+
+	var kcUser kc.UserRepresentation
+	kcUser, err = c.keycloakClient.GetUser(accessToken, realmName, *existingUser.UserID)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't get user from keycloak", "err", err.Error())
+		return errorhandler.CreateInternalServerError("keycloak")
+	}
+
+	if kcUser.Enabled == nil || *kcUser.Enabled {
+		c.logger.Warn(ctx, "msg", "Attempt to register a user with email of an already enabled user")
+		// Should not leak that email is already in use
+		return errorhandler.CreateBadRequestError(errorhandler.MsgErrInvalidParam + ".user_emailAddress")
+	}
+
+	err = c.keycloakClient.DeleteUser(accessToken, realmName, *existingUser.UserID)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't delete user from keycloak", "err", err.Error())
+		return errorhandler.CreateInternalServerError("keycloak")
+	}
+
+	err = c.usersDBModule.DeleteUser(ctx, realmName, *existingUser.UserID)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't delete user from database", "err", err.Error(), "uid", *existingUser.UserID)
+		return errorhandler.CreateInternalServerError("database")
+	}
+
+	// Free to go on processing this user creation
+	return nil
+}
+
+func (c *component) generateUnusedUsername(ctx context.Context, realmName string) (string, error) {
+	for {
```

Comment:
Should we have a maximum number of retry ? If it needs more than 10 tries it surely means we have an issue somewhere...

Reply:
Or it means you a very huge number of users !